crates/rbuilder/src/provider/ipc_state_provider.rs

```diff
@@ -0,0 +1,580 @@
+    /// Crates new IPC Provider Factory by establishing connection to IPC given path to the IPC
+    /// req_timeout is the same across all requests to the IPC
+    pub fn new(ipc_path: &Path, req_timeout: Duration) -> Self {
+        let ipc_provider = RpcProvider::try_connect(ipc_path, req_timeout.into())
+            // there is no need to gracefully handle (or propagate) this error, if we cannot connect
+            // to IPC, then rbuilder cannot work
+            .expect("can't connect to IPC");
+
+        Self {
+            ipc_provider,
+            code_cache: Arc::new(DashMap::new()),
+            state_provider_by_hash: Arc::new(Cache::new(DEFAULT_STATE_CACHE_SIZE)),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(default, deny_unknown_fields)]
+pub struct IpcProviderConfig {
+    pub(crate) request_timeout_ms: u64,
+    pub(crate) ipc_path: PathBuf,
+    pub(crate) mempool_server_url: String,
+}
+
+impl Default for IpcProviderConfig {
+    fn default() -> Self {
+        Self {
+            request_timeout_ms: DEFAULT_IPC_REQUEST_TIMEOUT_MS,
+            mempool_server_url: String::new(),
+            ipc_path: PathBuf::new(),
+        }
+    }
+}
+
+impl StateProviderFactory for IpcStateProviderFactory {
+    /// Gets state for the latest block
+    fn latest(&self) -> ProviderResult<StateProviderBox> {
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            BlockNumberOrTag::Latest.into(),
+            self.code_cache.clone(),
+        );
+
+        Ok(state)
+    }
+
+    /// Gets state at the block number
+    // We are not caching state provider by block number to avoid any issues with reorgs
+    // The calls to  history_by_block_number are rare and we shouldn't be loosing perf by not
+    // leveraging caching here
+    fn history_by_block_number(&self, block: BlockNumber) -> ProviderResult<StateProviderBox> {
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            block.into(),
+            self.code_cache.clone(),
+        );
+
+        Ok(state)
+    }
+
+    /// Gets state at the block hash
+    fn history_by_block_hash(&self, block: BlockHash) -> ProviderResult<StateProviderBox> {
+        if let Some(state) = self.state_provider_by_hash.get(&block) {
+            return Ok(Box::new(state));
+        }
+
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            block.into(),
+            self.code_cache.clone(),
+        );
+
+        self.state_provider_by_hash.insert(block, *state.clone());
+        Ok(state)
+    }
+
+    /// Gets block header given block hash
+    fn header(&self, block_hash: &BlockHash) -> ProviderResult<Option<Header>> {
+        let span =
+            trace_span!("header", id = rand::random::<u64>(), block_hash = %block_hash.to_string());
+        let _guard = span.enter();
+        trace!("header: get");
```

Comment:
Consider using tracing instrument attribute feature to reduce boilerplate
See https://docs.rs/tracing/latest/tracing/attr.instrument.html